Bluetooth: audio: Add Hearing Access Service client and server

MariuszSkamra opened the pull request:
This commit introduces Hearing Access Service.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

szymon-czapracki commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,672 @@
+			       BT_GATT_PERM_WRITE_ENCRYPT, NULL, write_control_point, &has_server),
+	BT_GATT_CHARACTERISTIC(BT_UUID_HAS_ACTIVE_PRESET_INDEX,
+			       BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ_ENCRYPT,
+			       read_active_preset_index, NULL, &has_server), );
```
Extra space and comma before end parenthesis

MariuszSkamra replied:
done

szymon-czapracki commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,672 @@
+		return NULL;
+	}
+
+	SYS_DLIST_FOR_EACH_CONTAINER (&server->preset_list, preset, node) {
```
Extra space after function name

szymon-czapracki commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,672 @@
+		return BT_GATT_ERR(BT_ATT_ERR_UNLIKELY);
+	}
+
+	SYS_DLIST_FOR_EACH_CONTAINER (&server->preset_list, preset, node) {
```
ditto (extra space)

szymon-czapracki commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,672 @@
+	}
+
+	/* Preset record list shall always have its entries in ascending order */
+	SYS_DLIST_FOR_EACH_CONTAINER (&server->preset_list, tmp, node) {
```
ditto (extra space)

szymon-czapracki commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,585 @@
+	client->params.write.data = data;
+	client->params.write.length = length;
+
+	err = bt_gatt_write(client->conn, &client->params.write);
```
This will return 0 on success. (consider below if statement)

szymon-czapracki commented on tests/bluetooth/bsim_bt/compile.sh:
```diff
@@ -20,29 +20,4 @@ mkdir -p ${WORK_DIR}
 
 source ${ZEPHYR_BASE}/tests/bluetooth/bsim_bt/compile.source
 
-app=tests/bluetooth/bsim_bt/bsim_test_app conf_file=prj_split.conf \
```
I don't think this should be erased - for testing only audio we can comment out the rest if need be.

MariuszSkamra replied:
Yes, I've pushed it by mistake

MariuszSkamra replied:
done

szymon-czapracki committed:
Bluetooth: audio: Add Hearing Access Service shell

Thist adds shell for HAS

Signed-off-by: Szymon Czapracki <szymon.czapracki@codecoup.pl>

jhedberg commented:
Could you try to split this up into separate commits a bit more? Eg UUID additions, server, client, shell, bsim tests, etc

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+		return BT_GATT_ERR(BT_ATT_ERR_UNLIKELY);
+	}
+
+	if (offset > 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len < sizeof(*hdr)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
```
Consider moving these checks up before `client_find` - No need to find a client in those cases

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+	}
+
+	/* Check encryption is enabled */
+	if (level < BT_SECURITY_L2) {
+		return;
+	}
```
So these checks if the security level goes from e.g. > L2 to < L2? Can that ever happen? Or is `security_changed` called with L0 or L1 in non-error cases?

MariuszSkamra replied:
As far as I remember the security level cannot decrease

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+			atomic_set_bit(client->flags, CLIENT_FLAG_ATT_MTU_VALID);
+		}
+
+		if (bt_gatt_is_subscribed(client->conn, &has_svc.attrs[4], BT_GATT_CCC_INDICATE)) {
+			atomic_set_bit(client->flags, CLIENT_FLAG_CP_IND_ENABLED);
+		}
+
+		if (bt_gatt_is_subscribed(client->conn, &has_svc.attrs[4], BT_GATT_CCC_NOTIFY)) {
+			atomic_set_bit(client->flags, CLIENT_FLAG_CP_NFY_ENABLED);
+		}
```
Could/should we set/clear these bits in the `cfg_changed` callbacks instead?

MariuszSkamra replied:
This is rather delayed reading of CCC configuration from the storage that shall be done on bt_conn->connected() callback. This is a workaround for https://github.com/zephyrproject-rtos/zephyr/issues/42829

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+#define BT_HAS_IS_LOCAL(_has) ((_has) == &has_local)
+#endif /* CONFIG_BT_HAS */
+
+int bt_has_register(struct bt_has_register_param *param, struct bt_has **out)
```
```suggestion
int bt_has_register(const struct bt_has_register_param *param, struct bt_has **out)
```

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+
+	*out = &has_local;
+
+	CHECKIF(param == NULL || param->ops == NULL || param->ops->active_set == NULL)
+	{
```
```suggestion
	CHECKIF(param == NULL || param->ops == NULL || param->ops->active_set == NULL) {
```

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+		struct has_preset *preset = &has_preset_list[i];
+
+		/* Sort the presets in order of increasing ID */
+		for (int j = 0; j < ARRAY_SIZE(param->preset_param); j++) {
```
```suggestion
		for (size_t j = 0; j < ARRAY_SIZE(param->preset_param); j++) {
```

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+		struct bt_has_preset_register_param *preset_param = NULL;
+		struct has_preset *preset = &has_preset_list[i];
+
+		/* Sort the presets in order of increasing ID */
```
Is this required? Do we gain anything from doing that? 

```suggestion
		/* Sort the presets in order of increasing index */
```

MariuszSkamra replied:
Yes, as the presets shall be sent to client in increasing order. So it's better to sort them on init

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+		preset->id = preset_param->idx;
+		preset->properties = preset_param->properties;
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+		strncpy(preset->name, preset_param->name, ARRAY_SIZE(preset->name) - 1);
```
See the previous UTF8 comment

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+#endif /* CONFIG_BT_HAS */
+}
+
+int bt_has_preset_active_get(struct bt_has *has)
```
If this function is for the client only, we can guard it with `CONFIG_BT_HAS_CLIENT`.

Thalley replied:
This applies to multiple functions.

Thalley replied:
Why is this, and others, client only functions in the first place? Seems useful enough for the server as well.

MariuszSkamra replied:
I would like to avoid situation where Bluetooth service is interface for getting Active Preset by other parts of the system.
If the Active Preset has to be exposed to the other parts of the system, then Application implements HAS should have a getter function IMO.

Thalley commented on subsys/bluetooth/audio/has_internal.h:
```diff
@@ -0,0 +1,153 @@
+ */
+
+/*
+ * Copyright (c) 2021 Codecoup
```
```suggestion
 * Copyright (c) 2021-2022 Codecoup
```
or
```suggestion
 * Copyright (c) 2022 Codecoup
```

Thalley commented on subsys/bluetooth/audio/has_internal.h:
```diff
@@ -0,0 +1,153 @@
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/uuid.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
I don't think this is necessary for internal header files

Thalley commented on subsys/bluetooth/shell/has.c:
```diff
@@ -0,0 +1,155 @@
+	int result;
+	struct bt_has_register_param register_param = {
+		.preset_param = {
+			{.idx = g_universal_idx,
+			.properties = BT_HAS_PROP_WRITABLE | BT_HAS_PROP_AVAILABLE,
+			.name = "Universal"},
```
```suggestion
			{
				.idx = g_universal_idx,
				.properties = BT_HAS_PROP_WRITABLE | BT_HAS_PROP_AVAILABLE,
				.name = "Universal"
			},
```
to keep the same style of `{`

Thalley commented on subsys/bluetooth/shell/has.c:
```diff
@@ -0,0 +1,155 @@
+static int cmd_has_set_active(const struct shell *sh, size_t argc, char **argv)
+{
+	int result;
+	int index = strtol(argv[1], NULL, 0);
```
```suggestion
	const int index = strtol(argv[1], NULL, 0);
```

Optional as this is "just" the shell. Feel free to do it, or skip it (for this and other similar uses).

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c:
```diff
@@ -0,0 +1,279 @@
+	.disconnected = disconnected,
+};
+
+static void bt_ready(int err)
+{
+	if (err != 0) {
+		FAIL("Bluetooth discover failed (err %d)\n", err);
+		return;
+	}
+
+	g_bt_init = true;
+}
```
You can omit the callback to `bt_enable` to make it synchronous and avoid this function and `g_bt_init`

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c:
```diff
@@ -0,0 +1,279 @@
+
+extern enum bst_result_t bst_result;
+
+static volatile bool g_bt_init;
+static volatile bool g_is_connected;
+static volatile bool g_mtu_exchanged;
+static volatile bool g_has_discovered;
+static volatile bool g_active_preset_notified;
+static volatile bool g_preset_read;
```
Consider using the CREATE_FLAG macros. Either should work, so this is fine too :)

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/has_test.c:
```diff
@@ -0,0 +1,161 @@
+	PASS("HAS passed\n");
+}
+
+static const struct bst_test_instance test_has[] = { { .test_id = "has_standalone",
+						       .test_post_init_f = test_init,
+						       .test_tick_f = test_tick,
+						       .test_main_f = test_standalone },
+						     { .test_id = "has",
+						       .test_post_init_f = test_init,
+						       .test_tick_f = test_tick,
+						       .test_main_f = test_main },
+						     BSTEST_END_MARKER };
```
```suggestion
static const struct bst_test_instance test_has[] = {
	{ 
		.test_id = "has_standalone",
		.test_post_init_f = test_init,
		.test_tick_f = test_tick,
		.test_main_f = test_standalone 
	},
	{ 
		.test_id = "has",
		.test_post_init_f = test_init,
		.test_tick_f = test_tick,
		.test_main_f = test_main 
	},
	BSTEST_END_MARKER 
};
```

Thalley commented on tests/bluetooth/shell/audio.conf:
```diff
@@ -101,3 +105,5 @@ CONFIG_BT_DEBUG_MEDIA_PROXY=y
 CONFIG_BT_DEBUG_PACS=y
 CONFIG_BT_AUDIO_DEBUG_UNICAST_SERVER=y
+
+CONFIG_BT_DEBUG_HAS=y
```
```suggestion
CONFIG_BT_DEBUG_HAS=y
CONFIG_BT_DEBUG_HAS_CLIENT=y
```

Thalley reviewed (changes requested):
Have not yet reviewed the HAS client, but this is a start. 

So far the structure and code quality looks great, and my comments are mostly minor code style comments. Good work!

asbjornsabo reviewed (commented):
Temporary comments - Git hub seems to be acting up, and I want to save the comments I have made so far.

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.has:
```diff
@@ -0,0 +1,89 @@
+	  This option enables support for Hearing Access Service Client.
+
+if BT_HAS
+choice BT_HAS_TYPE
```
Hearing Aid

asbjornsabo commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+
+#if defined(CONFIG_BT_HAS)
+static struct has_preset {
+	uint8_t id;
```
I am confused about the relation between "id" and "index".
Here, the preset has an "id", which, from elsewhere in the code, seems to be the same as the "index" of a preset.
Ditto for the `last_preset_id` in line 97.
But "index" also seems to be used for the same thing, e.g. in the `preset_foreach` function.  And in some calls to that function, "id"s are supplied where "index"es are specified as parameters.
It seems to me that these are the same, and then they should have the same name.  Preferably "index", as that is what is used in the spec.  (Although I think that is poor naming in the spec.)

To add to the confusion, there is also the "client_index", which seems to be an actual index used to index the array of clients.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c:
```diff
@@ -0,0 +1,279 @@
+		return;
+	}
+
+	k_sleep(K_MSEC(100));
```
Why are we waiting 100ms here?

MariuszSkamra replied:
Waiting for notification, fixed with condition WAIT_FOR

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/has_client_test.c:
```diff
@@ -0,0 +1,279 @@
+	PASS("HAS main PASS\n");
+}
+
+static const struct bst_test_instance test_has[] = { { .test_id = "has_client",
+						       .test_post_init_f = test_init,
+						       .test_tick_f = test_tick,
+						       .test_main_f = test_main },
+						     BSTEST_END_MARKER };
```
```suggestion
static const struct bst_test_instance test_has[] = { 
	{ 
		.test_id = "has_client",
		.test_post_init_f = test_init,
		.test_tick_f = test_tick,
		.test_main_f = test_main 
	},
	BSTEST_END_MARKER 
};
```

asbjornsabo commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+		return call_ops_set_active_preset(preset->id, sync);
+	}
+
+	preset_foreach(0x01, has_local.active_id - 1, 1, find_available, &preset);
```
Off-topic: This I have seen many times (elswhere).  Why are numbers some times given in hex, and sometimes not?  The index is given in hex - "0x01", while `num_matches` is given as "1", in decimal.

MariuszSkamra replied:
fixed to use decimal

asbjornsabo commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+	BT_DBG("conn %p attr %p offset %d", conn, attr, offset);
+
+	if (offset > 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_LONG);
```
Why that error code here, and "BT_ATT_ERR_INVALID_OFFSET" elswhere?

MariuszSkamra replied:
Fixed to use Invalid Offset error, however, Core 5.3 Vol 3, Part F, 3.4.4.5 says
"If the attribute value has a fixed length that is less than or equal to
(ATT_MTU - 1) octets in length, then an ATT_ERROR_RSP PDU may be sent
with the Error Code parameter set to Attribute Not Long (0x0B)."

So the behavior was correct

asbjornsabo commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+	case BT_HAS_OP_WRITE_PRESET_NAME:
+		if (!atomic_test_bit(client->flags, CLIENT_FLAG_CP_IND_ENABLED)) {
```
Because indications are mandatory for the HARC.
But is is mandatory for the server to ensure that the client is well-behaving?
Probably better to give error message, as is done here, than to accept and not get response through.

MariuszSkamra replied:
> Because indications are mandatory for the HARC. But is is mandatory for the server to ensure that the client is well-behaving? Probably better to give error message, as is done here, than to accept and not get response through.

HAS spec says:
"If a client writes 
    • The Read Presets Request opcode, or
    • The Write Preset Name opcode, or
    • The Set Active Preset opcode, or
    • The Set Active Preset – Synchronized Locally opcode
to the Hearing Aid Control Point characteristic, and the Client Characteristic Configuration Descriptor of the Hearing Aid Control Point is not configured for indications, the server shall return an ATT_ERROR_RSP PDU to the ATT_WRITE_REQ PDU with the error code parameter set to Client Characteristic Configuration Descriptor Improperly Configured as defined in [3]."

asbjornsabo reviewed (commented):
Not done, but need to call it a day.

I have several comments and questions, but the overall impression so far is good :-)

asbjornsabo reviewed (changes requested):
See comments

MariuszSkamra commented on include/bluetooth/audio/has.h:
```diff
@@ -0,0 +1,370 @@
+	/**
+	 * @brief Preset set active callback
+	 *
+	 * Once the preset becomes active, the bt_has_preset_active_set shall
+	 * be called to notify all the clients.
```
This is server only. The active_set callback is called then remote HAS Client attempts to change the preset. If the operation might be not possible to perform at the moment (Implementation specific) the application can return -EBUSY to simply reject the request.

But once the Preset becomes active, the application shall call `bt_has_preset_active_set`; the method used to inform all of the HAS Clients about currently used preset.

The Control Point operations is asynchronous, thus when Client sends Control Point ATT Write Request to Set Active Preset and gets ATT Write Response with error == 0 it cannot assume the active preset changed.

The specification says: "If this operation causes the value of the Active Preset Index characteristic to change, then the server shall notify clients of the new Active Preset Index value (see Section 3.3)."

So once the Server change the preset it will notify. But is something will fail - e.g. DSP re-configuration the notification will not be sent.

Thalley replied:
@MariuszSkamra The downside of this design is of course that HAS cannot be qualified without an application on top that is responsible for sending the notifications. This is the issue we discussed in the last F2F. 

w.r.t. my question: Would it be possible for HAS to send the notification autonomously if the return code is 0?

MariuszSkamra replied:
OK

MariuszSkamra commented on include/bluetooth/audio/has.h:
```diff
@@ -0,0 +1,370 @@
+ *
+ * @param has	Service object.
+ *
+ * @return Bluetooth connection object or NULL if local service object.
```
I've changed it to
int bt_has_client_conn_get(const struct bt_has *has, struct bt_conn **conn)

so that an error is returned

MariuszSkamra commented on include/bluetooth/audio/has.h:
```diff
@@ -0,0 +1,370 @@
+	/**
+	 * @brief Callback function for Hearing Access Service preset.
+	 *
+	 * Called when the value is read by client or changed by the remote server.
```
This one. The Change ID used for added presets and presets that name changed is the same "Generic ID"

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.has:
```diff
@@ -0,0 +1,89 @@
+
+if BT_AUDIO
+
+config BT_HAS
+	bool "Hearing Access Service support [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  This option enables support for Hearing Access Service.
```
Thanks, for that, apparently I forgot to look at security requirements. Those are as follows:
encryption key with at least 128 bits of entropy, derived from any of the following:
    • LE Secure Connections 
    • BR/EDR Secure Connections (if Cross-Transport Key Derivation is used)
    • OOB method 

so SC only mode is not required. I will update the implementation.

Regarding dynamic DB, I use static declaration.

MariuszSkamra commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+#include <bluetooth/gatt.h>
+#include <bluetooth/audio/has.h>
+
+#include "../host/conn_internal.h"
```
to have access to conn->state

Thalley replied:
I think I commented that somewhere: I think we can avoid doing that check in this layer, and let the GATT layer do that check.

MariuszSkamra commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+	return 0;
+}
+
+static void client_preset_read_cb(struct bt_conn *conn, uint8_t err,
+				  struct bt_gatt_write_params *write)
```
It's HAS Read Preset Request callback. The HAS Read Preset Request operation is actually ATT Write Request to Control Point. The HAS Read Preset Response is actually ATT Indication/Notification.

From HAS point of view it's Preset Read callback

MariuszSkamra commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+	return 0;
+}
+
+static void client_preset_read_cb(struct bt_conn *conn, uint8_t err,
```
Shall call preset_read_complete with an err

MariuszSkamra commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+	return bt_gatt_subscribe(client->conn, &client->active_preset_subscription);
+}
+
+static uint8_t disc_active_preset_idx_cb(struct bt_conn *conn, uint8_t err,
+					 struct bt_gatt_read_params *read, const void *data,
+					 uint16_t length)
```
Yes, but I'm OK to rename it. Done

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+#define LOG_MODULE_NAME bt_has
+#include "common/log.h"
+
+#define BT_HAS_MAX_CLIENT MIN(CONFIG_BT_MAX_CONN, CONFIG_BT_MAX_PAIRED)
```
The service is accessible for paired devices only, so it'll keep data for up to CONFIG_BT_MAX_PAIRED. The implementation keeps no context data for non-connected devices for memory efficiency.

So we maintain the paired and connected devices here only.

Thalley replied:
Should it not have been `CONFIG_BT_MAX_PAIRED` then?

e.g. in the case of `CONFIG_BT_MAX_PAIRED=5` and `CONFIG_BT_MAX_CONN=1`, then this will be 1.

MariuszSkamra replied:
It's as expected. Once the device is disconnected has_client is free'd.

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+
+#define BT_HAS_MAX_CLIENT MIN(CONFIG_BT_MAX_CONN, CONFIG_BT_MAX_PAIRED)
+#define BT_HAS_CP_WORK_TIMEOUT K_MSEC(10)
+#define BT_HAS_CP_TX_MAX_LEN                                                                       \
+	(1 + 2 /* Preset Changed */ + 3 /* Generic Update */ + BT_HAS_PRESET_NAME_MAX)
```
It's the maximum payload length used in HAS

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+
+	ARG_UNUSED(dev);
+
+	BUILD_ASSERT(BT_HAS_CP_TX_MAX_LEN < CONFIG_BT_L2CAP_TX_MTU);
```
Will fix it

asbjornsabo commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+	return BT_GATT_ITER_STOP;
+}
+
+static int cp_discover(struct bt_has_client *client)
```
Should this be `disc_cp`, for consistency with the discovery functions for the two other characteristics?

asbjornsabo commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+	return control_point_tx(client, client_preset_active_set_cb, buf, sizeof(buf));
+}
+
+int bt_has_discover(struct bt_conn *conn)
```
It would have helped me if the function had a few comment lines on the top describing the general approach:
- discover first characteristic, subscribe to it, read it, discover next characteristic, ...

Some of the other audio services has this (e.g. VCS, see below).  And as this service follows another pattern, an explanation here might also be in place.

(From vcs_client.c:)
	/*
	 * This will initiate a discover procedure. The procedure will do the
	 * following sequence:
	 * 1) Primary discover for the VCS
	 * 2) Characteristic discover of the VCS
	 * 3) Discover services included in VCS (VOCS and AICS)
	 * 4) For each included service found; discovery of the characteristics
	 * 5) When everything above have been discovered, the callback is called
	 */

asbjornsabo commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+	client->params.read.by_uuid.start_handle = 0x0001;
+	client->params.read.by_uuid.end_handle = 0xffff;
+
+	return bt_gatt_read(client->conn, &client->params.read);
```
Why does this discovery function use `bt_gatt_read` while the discovery functions for the two other characteristics use `bt_gatt_discover`?

MariuszSkamra replied:
Because it performs Read Using Characteristic UUID GATT operation. So doing 1 operation, the characteristic is read and discovered.

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+	BT_DBG("attr %p value 0x%04x", attr, value);
+}
+
+static ssize_t cp_ccc_cfg_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				uint16_t value)
```
I'll do something similar you suggested

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+static void control_point_ind_done(struct bt_conn *conn, struct bt_gatt_indicate_params *params,
+				   uint8_t err)
+{
+	/* TODO: handle error somehow */
```
Yes, but what to do with this error? Pass to application that remote device did not acked indication? :D I thought rather about retry

Thalley replied:
> Yes, but what to do with this error? Pass to application that remote device did not acked indication?

Yes, because then the application can determine if it wants to attempt again. 

> I thought rather about retry

Not a bad idea, but what if the client continuously rejects the indication? Would the application be better suited to determine what to do in such a case?

MariuszSkamra replied:
But the application does not even know that there was a operation ongoing - i.e. Read Presets operation is handled by the stack. I'll just put an error log here as you suggested for now

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+{
+	if (atomic_test_bit(client->flags, CLIENT_FLAG_CP_NFY_ENABLED)) {
+		client->cp_tx_params.nfy.attr = &has_svc.attrs[4];
+		client->cp_tx_params.nfy.func = control_point_tx_done;
```
It's because control_point_tx_done is common for IND and NFY. I have to rename NFY to NTF - that abbreviation is used in Core Spec

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+		client->cp_tx_params.nfy.data = buf->data;
+		client->cp_tx_params.nfy.len = buf->len;
+
+		return bt_gatt_indicate(client->conn, &client->cp_tx_params.ind);
```
Yes, I've already catched it on IOP

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+	switch (pc->change_id) {
+	case BT_HAS_CHANGE_ID_GENERIC_UPDATE: {
+		struct bt_has_cp_generic_update *gu;
+		size_t name_len = MIN(sizeof(preset->name) - 1, BT_HAS_PRESET_NAME_MAX);
```
Indeed, that was a bug  I've already fixed, thanks

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+
+typedef bool (*preset_func_t)(struct has_preset *preset, void *user_data);
+
+static void preset_foreach(uint8_t start_index, uint8_t end_index, uint8_t num_matches,
```
I'll leave it as I started to use it

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+		return BT_GATT_ERR(BT_HAS_ERR_WRITE_NAME_NOT_ALLOWED);
+	}
+
+	strncpy(preset->name, name, len);
```
truncate, fixed

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.has:
```diff
@@ -0,0 +1,89 @@
+
+endchoice # BT_HAS_HA_TYPE
+
+config BT_HAS_PRESET_CNT
+	int "Preset record list size"
+	default 4
+	range 0 255
+	help
+	  This option sets the number of Hearing Access Service Presets
+	  that can be registered.
```
I still need to know the maximum number of presets for internal purposes

Thalley replied:
Makes sense. Should the minimum value be 0 though? Should we not require that if you enable HAS support, that we support at least 1 preset?

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.has:
```diff
@@ -0,0 +1,89 @@
+	help
+	  Set if there is support for active preset changes relay to the other
+	  member in the Binaural Hearing Aid Set.
+
```
BT_HAS_WRITABLE_PRESETS is an property that is exposed to the client. As per spec, it tells whether the preset name can be changed by the client. Even if it's disabled, it can be still changed by Server so BT_HAS_PRESET_NAME_DYNAMIC is the server feature to set the name.

Presets that have BT_HAS_WRITABLE_PRESETS property not set can still change it's name when done by server itself. That is how I understand it.

It's just like in GATT, where:
BT_DEVICE_NAME_DYNAMIC shall be set + BT_DEVICE_NAME_GATT_WRITABLE shall be set to allow the GATT client to change the device name

MariuszSkamra replied:
Regarding BT_HAS_DYNAMIC_PRESETS I agree, if the support is disabled the bt_has_preset_visibility_set shall return an error. Will fix it.

asbjornsabo replied:
I am not following.
I see your point regarding the PRESET_NAME_DYNAMIC.  But how is the "writable preset support" field decided - is that by seeing if any of the preset names are writable?
And how is the "dynamic presets" field decided?

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+
+		return bt_gatt_indicate(client->conn, &client->cp_tx_params.ind);
+	}
+
```
Can do. I'll leave TODO here to determine which PDU to send

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+
+	switch (hdr->op) {
+	case BT_HAS_OP_READ_PRESET_REQ:
+		if (!atomic_test_bit(client->flags, CLIENT_FLAG_CP_IND_ENABLED)) {
```
Fixed to test the Notifications bit as well.

The Active Preset Index is separate characteristic. There is no statement in the specification that Server shall rejest the control point write when Client did not subscribe for Active Preset Index notifications

MariuszSkamra replied:
HAS specification says:
"
If a client writes 
    • The Read Presets Request opcode, or
    • The Write Preset Name opcode, or
    • The Set Active Preset opcode, or
    • The Set Active Preset – Synchronized Locally opcode
to the Hearing Aid Control Point characteristic, and the Client Characteristic Configuration Descriptor of the Hearing Aid Control Point is not configured for indications, the server shall return an ATT_ERROR_RSP PDU to the ATT_WRITE_REQ PDU with the error code parameter set to Client Characteristic Configuration Descriptor Improperly Configured as defined in [3].
"

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+
+	BUILD_ASSERT(BT_HAS_CP_TX_MAX_LEN < CONFIG_BT_L2CAP_TX_MTU);
+
+	if (IS_ENABLED(CONFIG_BT_HAS_HA_TYPE_MONAURAL)) {
+		type = BT_HAS_MONAURAL_HEARING_AID;
+	} else if (IS_ENABLED(CONFIG_BT_HAS_HA_TYPE_BANDED)) {
+		type = BT_HAS_BANDED_HEARING_AID;
+	} else {
+		type = BT_HAS_BINAURAL_HEARING_AID;
+	}
```
Kconfig's choice cannot store value AFAIK

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+			break;
+		}
+
+		preset->id = preset_param->idx;
```
those are the same. ID is leftover, I renamed it to idx

Thalley commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+	       client->read_preset_rsp.num_presets > 0;
+}
+
+static void control_point_tx_work_submit(struct has_client *client, k_timeout_t delay)
```
Correct, but you use the same naming scheme for the client, where in the context of the client it refers to sending a control point operation. 

I don't feel using the same name for the server and client, when they are doing two different operation, is a good choice. I am tempted to suggest `control_point_rsp_work_submit` instead (or `cp_rsp_work_submit`), but `rsp` is not really true to the operation either, as it will be a notification or indication, and not really a response (although given the design of HAS, it acts as a response).

MariuszSkamra replied:
I renamed it to control_point_write for client (as it uses ATT Write Request) and control_point_send for server (notification/indication)

MariuszSkamra commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+	client->busy = true;
+	client->conn = conn;
+
+	err = disc_hearing_aid_features(client);
```
As there can be only one instance of HAS, it's faster to discover the characteristics (all are mandatory).
Discovery of the service first is redundant.

MariuszSkamra commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+{
+	client->busy = false;
+
+	has_cb->discover(client->conn, success ? &client->has : NULL,
```
added guards

MariuszSkamra commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+					struct bt_gatt_subscribe_params *params, const void *data,
+					uint16_t length)
+{
+	struct bt_has_client *client = client_by_conn(conn);
```
Will fix that. However, I saw that the conn is not guarded in other audio related services, e.g will result in assert in csis_client.c, so this is a common issue

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+			break;
+		}
+
+		index += 32;
```
By mistake, I should be like this:
u32_count_trailing_zeros(atomic_get(&client->preset_changed.pending[i])) + i * 32;
The reason is I traverse atomic_t array looking for the first bit set.

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+#include <bluetooth/gatt.h>
+
+#include "has_internal.h"
+#include "../host/conn_internal.h"
+#include "../host/hci_core.h"
```
I need to use bt_addr_le_is_bonded which is non-public API

MariuszSkamra commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+	client->active_preset_subscription.value_handle = handle;
+	client->active_preset_subscription.ccc_handle = 0x0000;
+	client->active_preset_subscription.end_handle = 0xffff;
+	client->active_preset_subscription.disc_params = &client->params.discover;
```
I do not attempt to discover multiple values at the same time I guess. Once discovery of attribute is done I use the same parameters to discover another one

MariuszSkamra commented on subsys/bluetooth/audio/has.c:
```diff
@@ -0,0 +1,1330 @@
+#else
+	const char *name;
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+	bool hidden;
```
No, it's rather to hide the presets from the client. It will result in Preset Deleted change ID notification.

MariuszSkamra committed:
Bluetooth: uuid: Add Hearing Access Service

This adds Hearing Access Service UUID values.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: Add Hearing Access Service

This adds Hearing Access Service client and server roles.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: tests: Add Hearing Access Service bsim tests

This extends the bsim tests with the HAS client/server and server
standalone test cases.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented on subsys/bluetooth/audio/has_client.c:
```diff
@@ -0,0 +1,770 @@
+		BT_INFO("HAS Control Point not found");
+		disc_complete(client, true);
+		return BT_GATT_ITER_STOP;
+	}
```
You have right, I will need to pass this information to the app somehow.

asbjornsabo commented on include/bluetooth/audio/has.h:
```diff
@@ -0,0 +1,350 @@
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_has_preset_visibility_set(struct bt_has *has, uint8_t index, bool visible);
```
What is the purpose and reason for (adding) this concept of "visibility" for presets?  It does not seem to be a feature of the specification.

asbjornsabo commented on include/bluetooth/audio/has.h:
```diff
@@ -0,0 +1,350 @@
+	 *  @return 0 in case of success or negative value in case of error.
+	 *  @return -EBUSY if operation cannot be executed at the time.
+	 *  @return -EINPROGRESS in case where user has to confirm once the requested preset
+	 *                       becomes active by calling @ref bt_has_preset_active_set.
```
Not fully clear to me.  Does the preset become active by calling `bt_has_preset_active_set`?  Or does the user confirm by calling `bt_has_preset_active_set`?
How does the user confirm?  And does that happen before or after the call to `bt_has_preset_active_set`?
